pkg/server/plugin/nodeattestor/httpchallenge/httpchallenge_test.go

```diff
@@ -0,0 +1,433 @@
+	}{
+		{
+			name:        "Attest succeeds for defaults",
+			hclConf:     "",
+			tofu:        true,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+			expectedAgentID: "spiffe://example.org/spire/agent/http_challenge/foo",
+			expectedSelectors: []*common.Selector{
+				{
+					Type:  "http_challenge",
+					Value: "hostname:foo",
+				},
+			},
+		},
+		{
+			name:        "Attest succeeds for reattest without tofu",
+			hclConf:     "tofu = false\nallow_non_root_ports = false",
+			tofu:        false,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+			expectedAgentID: "spiffe://example.org/spire/agent/http_challenge/foo",
+			expectedSelectors: []*common.Selector{
+				{
+					Type:  "http_challenge",
+					Value: "hostname:foo",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			testNonce := "123456789abcdefghijklmnopqrstuvwxyz"
+			plugin := loadPlugin(t, tt.hclConf, !tt.tofu, client, testNonce)
+			result, err := plugin.Attest(context.Background(), tt.payload, tt.challengeFn)
+			require.NoError(t, err)
+			require.NotNil(t, result)
+
+			require.Equal(t, tt.expectedAgentID, result.AgentID)
+			requireSelectorsMatch(t, tt.expectedSelectors, result.Selectors)
+		})
+	}
+}
+
+func loadPlugin(t *testing.T, config string, testTOFU bool, client *http.Client, testNonce string) nodeattestor.NodeAttestor {
+	v1 := new(nodeattestor.V1)
+	agentStore := fakeagentstore.New()
+	var configureErr error
+	if testTOFU {
+		agentStore.SetAgentInfo(&agentstorev1.AgentInfo{
+			AgentId: "spiffe://example.org/spire/agent/http_challenge/foo",
+		})
+	}
+	opts := []plugintest.Option{
+		plugintest.Configure(config),
+		plugintest.CaptureConfigureError(&configureErr),
+		plugintest.HostServices(agentstorev1.AgentStoreServiceServer(agentStore)),
+		plugintest.CoreConfig(catalog.CoreConfig{
+			TrustDomain: spiffeid.RequireTrustDomainFromString("example.org"),
+		}),
+	}
+	plugintest.Load(t, httpchallenge.BuiltInTesting(client, testNonce), v1, opts...)
+	return v1
+}
+
+func marshalPayload(t *testing.T, attReq *common_httpchallenge.AttestationData) []byte {
+	attReqBytes, err := json.Marshal(attReq)
+	require.NoError(t, err)
+	return attReqBytes
+}
+
+func requireSelectorsMatch(t *testing.T, expected []*common.Selector, actual []*common.Selector) {
+	require.Equal(t, len(expected), len(actual))
+	for idx, expSel := range expected {
+		require.Equal(t, expSel.Type, actual[idx].Type)
+		require.Equal(t, expSel.Value, actual[idx].Value)
+	}
+}
+
+func newClientWithLocalIntercept(url string) *http.Client {
+	u, _ := neturl.Parse(url)
+	_, port, _ := net.SplitHostPort(u.Host)
+	return &http.Client{
+		Transport: &http.Transport{
```

Comment:
Another option here is to clone the default transport like in my other comment and then replace the dialer.